zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,498 @@
+            let input = meta.query_advice(input, Rotation::cur());
+            let code = decode::data_expr(&parts);
+            let word_begin = meta.query_advice(word_begin, Rotation::cur());
+            vec![
+                (word_begin.clone() * input, huffman_table[0]),
+                (word_begin.clone() * code, huffman_table[1]),
+                (word_begin * decode::num_bits_expr(&parts), huffman_table[2]),
+            ]
+        });
+
+        meta.create_gate("do RLC over parts", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+            cb.condition(meta.query_advice(word_begin, Rotation::cur()), |cb| {
+                let mut data_rlc_prev = meta.query_advice(data_rlc, Rotation::prev());
+                for part in parts.iter() {
+                    // 1. TODO: range check(lookup):
+                    //      part.data(0..2^8)
+                    //      part.num_bits(0..32)
+                    //      part.r_multi(bool)
+                    //      part.shift_factor(0..2^32) 2.pow(0..32)
+                    //      part.decode_factor(0..2^32) 2.pow(0..32)
+                    cb.require_boolean("r_multi", part.r_multi.clone());
+                    // 2. Add the part to the rlc
+                    let multi = select::expr(part.r_multi.clone(), randomness.clone(), 1.expr());
+                    data_rlc_prev =
+                        data_rlc_prev * multi + part.data.clone() * part.shift_factor.clone();
+                }
+                cb.require_equal(
+                    "data_rlc",
+                    data_rlc_prev,
+                    meta.query_advice(data_rlc, Rotation::cur()),
+                );
+            });
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        Self {
+            q_enable,
+            word_begin,
+            input,
+            data,
+            data_rlc,
+            num_bits,
+            r_multi,
+            shift_factor,
+            decode_factor,
+            huffman_table,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<F: Field> CompressCircuitConfig<F> {
+    fn load_huffman_table(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        layouter.assign_table(
+            || "fixed huffman table",
+            |mut table| {
+                table.assign_cell(
+                    || "huffman table value",
+                    self.huffman_table[0],
+                    0,
+                    || Value::known(F::zero()),
+                )?;
+                table.assign_cell(
+                    || "huffman table code",
+                    self.huffman_table[1],
+                    0,
+                    || Value::known(F::zero()),
+                )?;
+                table.assign_cell(
+                    || "huffman table code length",
+                    self.huffman_table[2],
+                    0,
+                    || Value::known(F::zero()),
+                )?;
```

Comment:
Instead of doing these 3 separate assignments, a bit cleaner to just use a loop over `self.huffman_table.iter()`. Perhaps the same is also true for the non-zero assignments below.

Reply:
The zero cells just be used for non-word-begin rows